test/integration/job/job_test.go

```diff
@@ -1174,6 +1174,263 @@ func TestBackoffLimitPerIndex(t *testing.T) {
+	podTemplateSpec := v1.PodTemplateSpec{
+		Spec: v1.PodSpec{
+			Containers: []v1.Container{
+				{
+					Name:                     "main-container",
+					Image:                    "foo",
+					ImagePullPolicy:          v1.PullIfNotPresent,
+					TerminationMessagePolicy: v1.TerminationMessageFallbackToLogsOnError,
+				},
+			},
+		},
+	}
+
+	resetMetrics()
+	defer featuregatetesting.SetFeatureGateDuringTest(t, feature.DefaultFeatureGate, features.JobManagedByLabel, true)()
+
+	closeFn, restConfig, clientSet, ns := setup(t, "managed-by-recreate-job")
+	defer closeFn()
+	ctx, cancel := startJobControllerAndWaitForCaches(t, restConfig)
+	defer func() {
+		cancel()
+	}()
+	baseJob := batchv1.Job{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "custom-job-test",
+			Namespace: ns.Name,
+			Labels:    make(map[string]string),
+		},
+		Spec: batchv1.JobSpec{
+			Completions: ptr.To[int32](1),
+			Parallelism: ptr.To[int32](1),
+			Template:    podTemplateSpec,
+		},
+	}
+	jobObj, err := createJobWithDefaults(ctx, clientSet, ns.Name, &baseJob)
+	if err != nil {
+		t.Fatalf("Error %q while creating the job %q", err, jobObj.Name)
+	}
+	validateJobPodsStatus(ctx, t, clientSet, jobObj, podsByStatus{
+		Active:      1,
+		Ready:       ptr.To[int32](0),
+		Terminating: ptr.To[int32](0),
+	})
+	if err, _ := setJobPodsPhase(ctx, clientSet, jobObj, v1.PodSucceeded, 1); err != nil {
+		t.Fatalf("Failed setting phase %s on Job Pod: %v", v1.PodFailed, err)
+	}
+
+	validateJobSucceeded(ctx, t, clientSet, jobObj)
+
+	err = clientSet.CoreV1().Pods(jobObj.Namespace).DeleteCollection(ctx, metav1.DeleteOptions{}, metav1.ListOptions{})
+	if err != nil {
+		t.Fatalf("Error: '%v' while deleting the pods", err)
+	}
+
+	jobClient := clientSet.BatchV1().Jobs(jobObj.Namespace)
+	err = jobClient.Delete(ctx, jobObj.Name, metav1.DeleteOptions{})
+	if err != nil {
+		t.Fatalf("Failed to delete job: %v", err)
+	}
+	jobObj, err = jobClient.Get(ctx, jobObj.Name, metav1.GetOptions{})
+	if err != nil {
+		t.Fatalf("Failed to get deleted job: %v", err)
+	}
+
+	jobObj, err = updateJob(ctx, jobClient, jobObj.Name, func(j *batchv1.Job) {
+		j.Finalizers = []string{}
+	})
+	if err != nil {
+		t.Fatalf("Failed to clean up the orphan finalizer: %v", err)
+	}
+
+	// Wait for pods to start up.
+	err = wait.PollUntilContextTimeout(ctx, 5*time.Millisecond, wait.ForeverTestTimeout, true, func(c context.Context) (done bool, err error) {
+		_, e := jobClient.Get(c, jobObj.Name, metav1.GetOptions{})
+		if apierrors.IsNotFound(e) {
+			return true, nil
+		}
+		return false, e
+	})
+	if err != nil {
+		t.Fatalf("waiting for job %#v to be deleted: %v", jobObj, err)
+	}
+
+	jobWithManagedBy := baseJob.DeepCopy()
+	jobWithManagedBy.Labels[batchv1.JobManagedByLabel] = "custom-job-controller"
+	jobObj, err = createJobWithDefaults(ctx, clientSet, ns.Name, jobWithManagedBy)
+	if err != nil {
+		t.Fatalf("Error %q while creating the job %q", err, jobObj.Name)
+	}
+
+	time.Sleep(2 * time.Second)
```

Comment:
Could we use an eventually or anything else other than a sleep here?

Reply:
We want to ensure there was not update to the job, so I don't see how to do it otherwise. 
 
`Eventually` is a function of gomega framework, I don't think it is used here, but we rather want to use `Consistently` here.

I don't see anything better than sleep, but maybe 2s is excessive. 

I'm thinking it might be good to make the worker interval configurable (it is 1s now, here: https://github.com/kubernetes/kubernetes/blob/47f92ce2311bc2e0b6907a23d45da5599e5fa84c/pkg/controller/job/job_controller.go#L232), and reduce to 100ms. Then, sleeping for 300ms should be enough.

Reply:
I see. I guess it would only get slower on high ci load.

Reply:
wait for the metric to be updated, then check the status once.